### ISSUE TITLE
As an API

### ISSUE THREAD

**Issue by JKSnd** [NickeManarin/ScreenToGif#33]
This could be a very powerful support tool if it could be included in a software project.

The idea as a workflow:

1.  A user of my software is having an issue
2. They hit the "Report Issue" button
3. Behind the scenes, screentogif.dll starts recording the application's window.
4. The user is instructed to reproduce the issue if possible
5. Once the user has recreated the problem, they indicate that they're done
6. The application stops the recording
7. The application acquires the generated gif
8. Logs are packaged up with the gif, it is all sent back to our support department as a tidy little ticket.

If such a library as screentogif.dll existed, and supported steps 3,4,6, and 7 then that would be stellar.

I'm sure we could pull the source and just do it in-house as a one-off, but everybody benefits when packages make it up to NuGet.

Thanks for considering!

**Comment by NickeManarin** [comment]
This is an awesome idea!

**Comment by Elusive138** [comment]
If you do, please consider also dual-licensing with an additional licence, because [MS-PL might not be GPL-compatible](https://softwareengineering.stackexchange.com/questions/191123/why-isnt-the-microsoft-public-license-compatible-with-the-gpl). Depending on how permissive you want to get, perhaps (one of) the LGPL or MIT licences.

**Comment by NickeManarin** [comment]
Yes, I want to change the license of this project, but I have to read everything carefully.

**Comment by Elusive138** [comment]
Thanks! :)

I suggested dual-licensing instead of changing the existing one so existing users won't have to reevaluate the licence. Might make things easier.

**Comment by NickeManarin** [comment]
Oh, I was thinking about switching to a license very close to the current one with the same reason in mind. :)

**Comment by JKSnd** [comment]
That would be awesome! Especially since I'm not a pro at interpreting 3d of the Ms-Pl. What constitutes a compatible license? I'm not sure.

I'm a big fan of the MIT, for what it's worth. 

Dual-licensing as a temporary way to smoothly transition may be a good option.

**Comment by vatterspun** [comment]
> ... please consider also dual-licensing with an additional licence, because MS-PL might not be GPL-compatible.

To clarify, according to the GNU.ORG site, MS-PL is [incompatible with the GPL](https://www.gnu.org/licenses/license-list.html#ms-pl).

> Yes, I want to change the license of this project, but I have to read everything carefully

Sounds great.  Whatever you choose, make sure it includes something about patents (something the MS-Pl definitely gets right).  So for example, Apache 2.0 does so while BSD and MIT do not.

I'm a fan of the GPL for a few reasons, but maybe the major one is that I like the idea of programs and their code living beyond their creators, where many "permissive" licenses can be taken in a fully closed source direction, as most notably happened in the case of Apple's Darwin and the Microsoft Windows networking stack.

ScreenToGif is one of the most active and interesting I've seen in a long time (and I've played with a LOT of animated GIF editors), so please don't let anyone influence you to select anything that will diminish your interest.

**Comment by Konard** [comment]
There is another way to approach the same issue. We can record everything at the background for the last hour or two, and when something unexpected happens, we can save a recording. This way we save user time, and there is no need to reproduce anything. This approach is similar to how the [black box](https://en.wikipedia.org/wiki/Flight_recorder) works.

**Comment by NickeManarin** [comment]
@Konard With the next version, there will be available a new capture mode, which only captures a frame when something actually changes within the capture region.

With this new feature, it will be possible to make long captures using less disk space.

**Comment by sharwell** [comment]
Hi @NickeManarin,

We are very interested in using a library version of ScreenToGif as part of our product integration test pipeline. It sounds like you may be interested in having this feature be part of ScreenToGif itself, so I wanted to discuss this possibility with you and determine if it would be best to try and create the library separately (derived from this project) or work on it here as a new product feature for everyone. Looking forward to hearing your preference.

As a side note, I have assisted other projects in license migration. If you are interested in migrating to another license, I'd be happy to explain the process we followed.

Thanks,
Sam
